File: app/financials/revenue_analytics_screen.py
import datetime
import random

class RevenueAnalyticsScreen:
    def __init__(self):
        self.revenue_data = {}
        self._generate_mock_revenue_data()

    def _generate_mock_revenue_data(self):
        # Mock data for revenue analytics
        today = datetime.date.today()

        # Revenue by source (monthly and YTD)
        sources = ["Consultation Fees", "Marketplace Commissions", "Subscription Fees", "Premium Content Sales"]
        revenue_by_source_mtd = {source: round(random.uniform(1000, 10000), 2) for source in sources}
        revenue_by_source_ytd = {source: val * random.uniform(8,12) for source, val in revenue_by_source_mtd.items()} # Approximate YTD

        total_revenue_mtd = sum(revenue_by_source_mtd.values())
        total_revenue_ytd = sum(revenue_by_source_ytd.values())

        # Revenue trends (e.g., daily for last 30 days, monthly for last 12 months)
        daily_revenue_trend = []
        for i in range(30):
            day = today - datetime.timedelta(days=i)
            daily_rev = sum(round(random.uniform(50, 200), 2) for _ in sources) # Simplified daily sum
            daily_revenue_trend.append({"date": day.isoformat(), "revenue": daily_rev})
        daily_revenue_trend.reverse()

        monthly_revenue_trend = []
        current_month = today.replace(day=1)
        for i in range(12):
            month_start = (current_month - datetime.timedelta(days=i*30)).replace(day=1) # Approximation
            monthly_rev = sum(round(random.uniform(5000, 20000), 2) for _ in sources) # Simplified monthly sum
            monthly_revenue_trend.append({"month": month_start.strftime("%Y-%m"), "revenue": monthly_rev})
        monthly_revenue_trend.reverse()

        # Customer Lifetime Value (CLV) - simplified
        clv_segments = {
            "farmers_standard": round(random.uniform(50, 150), 2),
            "farmers_premium": round(random.uniform(150, 400), 2),
            "veterinarians_basic": round(random.uniform(100, 250), 2),
            "veterinarians_pro": round(random.uniform(250, 600), 2),
            "overall_average": round(sum([random.uniform(50,600) for _ in range(4)])/4, 2)
        }

        # Churn Rate (mock)
        churn_rate_monthly = f"{random.uniform(1.5, 5.0):.2f}%"

        self.revenue_data = {
            "summary": {
                "total_revenue_mtd": total_revenue_mtd,
                "total_revenue_ytd": total_revenue_ytd,
                "average_revenue_per_paying_user_arppu_monthly": round(total_revenue_mtd / (1000 + random.randint(-100,100)),2), # Mock paying users
                "churn_rate_monthly": churn_rate_monthly,
            },
            "revenue_by_source": {
                "mtd": revenue_by_source_mtd,
                "ytd": revenue_by_source_ytd,
            },
            "revenue_trends": {
                "daily_last_30_days": daily_revenue_trend,
                "monthly_last_12_months": monthly_revenue_trend,
            },
            "customer_lifetime_value_clv": clv_segments,
            "top_revenue_generating_items": { # e.g. specific consultations, products
                "consult_specialist_A": {"type": "Consultation", "revenue_mtd": round(random.uniform(500,1500),2)},
                "product_feed_B": {"type": "Marketplace", "revenue_mtd": round(random.uniform(300,1000),2)},
                "subscription_vet_pro": {"type": "Subscription", "revenue_mtd": round(random.uniform(1000,2000),2)},
            }
        }

    def get_revenue_summary(self):
        return self.revenue_data.get("summary")

    def get_revenue_by_source(self, period="mtd"): # period can be 'mtd' or 'ytd'
        return self.revenue_data.get("revenue_by_source", {}).get(period)

    def get_revenue_trend(self, granularity="daily"): # granularity can be 'daily' or 'monthly'
        if granularity == "daily":
            return self.revenue_data.get("revenue_trends", {}).get("daily_last_30_days")
        elif granularity == "monthly":
            return self.revenue_data.get("revenue_trends", {}).get("monthly_last_12_months")
        return []

    def get_clv_data(self):
        return self.revenue_data.get("customer_lifetime_value_clv")

    def display_screen(self):
        print("---- Revenue Analytics Screen ----")

        summary = self.get_revenue_summary()
        if summary:
            print("\n-- Revenue Summary --")
            for key, value in summary.items():
                print(f"  {key.replace('_', ' ').title()}: ${value:,.2f}" if isinstance(value, float) else f"  {key.replace('_', ' ').title()}: {value}")

        print("\n-- Revenue by Source (MTD) --")
        source_mtd = self.get_revenue_by_source("mtd")
        if source_mtd:
            for source, amount in source_mtd.items():
                print(f"  {source}: ${amount:,.2f}")

        print("\n-- Revenue by Source (YTD) --")
        source_ytd = self.get_revenue_by_source("ytd")
        if source_ytd:
            for source, amount in source_ytd.items():
                print(f"  {source}: ${amount:,.2f}")

        print("\n-- Daily Revenue Trend (Last 5 Days) --")
        daily_trend = self.get_revenue_trend("daily")
        if daily_trend:
            for item in daily_trend[-5:]: # Last 5 days for brevity
                print(f"  Date: {item['date']}, Revenue: ${item['revenue']:,.2f}")

        print("\n-- Monthly Revenue Trend (Last 3 Months) --")
        monthly_trend = self.get_revenue_trend("monthly")
        if monthly_trend:
            for item in monthly_trend[-3:]: # Last 3 months for brevity
                print(f"  Month: {item['month']}, Revenue: ${item['revenue']:,.2f}")

        print("\n-- Customer Lifetime Value (CLV) by Segment --")
        clv_data = self.get_clv_data()
        if clv_data:
            for segment, value in clv_data.items():
                 print(f"  {segment.replace('_', ' ').title()}: ${value:,.2f}")

        print("\n-- Top Revenue Generating Items (MTD) --")
        top_items = self.revenue_data.get("top_revenue_generating_items", {})
        if top_items:
            for item, data in top_items.items():
                print(f"  Item: {item.replace('_', ' ').title()} ({data['type']}), Revenue: ${data['revenue_mtd']:,.2f}")

        print("---------------------------------")

if __name__ == '__main__':
    screen = RevenueAnalyticsScreen()
    screen.display_screen()

File: app/src/main/java/com/example/rooster/financials/revenueanalytics/RevenueAnalyticsScreen.kt
package com.example.rooster.financials.revenueanalytics

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.material3.*
import androidx.compose.runtime.Composable
import androidx.compose.runtime.collectAsState
import androidx.compose.runtime.getValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import androidx.lifecycle.ViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import java.text.DecimalFormat
import java.util.Calendar
import kotlin.random.Random

// --- Data Classes ---
// Using Map<String, Any> for flexibility, mirroring Python's dictionary structure.
// For a production app, more specific data classes would be better.
data class RevenueSummary(val metrics: Map<String, Any>)
data class RevenueBySource(val periodData: Map<String, Map<String, Double>>) // MTD/YTD -> Source -> Amount
data class RevenueTrendPoint(val label: String, val revenue: Double) // Date or Month string
data class RevenueTrend(val granularity: String, val points: List<RevenueTrendPoint>)
data class ClvData(val segments: Map<String, Double>)
data class TopRevenueItem(val name: String, val type: String, val revenueMtd: Double)

// --- ViewModel ---
class RevenueAnalyticsViewModel : ViewModel() {
    private val _summary = MutableStateFlow<RevenueSummary?>(null)
    val summary: StateFlow<RevenueSummary?> = _summary

    private val _revenueBySource = MutableStateFlow<RevenueBySource?>(null)
    val revenueBySource: StateFlow<RevenueBySource?> = _revenueBySource

    private val _dailyTrend = MutableStateFlow<RevenueTrend?>(null)
    val dailyTrend: StateFlow<RevenueTrend?> = _dailyTrend

    private val _monthlyTrend = MutableStateFlow<RevenueTrend?>(null)
    val monthlyTrend: StateFlow<RevenueTrend?> = _monthlyTrend

    private val _clv = MutableStateFlow<ClvData?>(null)
    val clv: StateFlow<ClvData?> = _clv

    private val _topItems = MutableStateFlow<List<TopRevenueItem>>(emptyList())
    val topItems: StateFlow<List<TopRevenueItem>> = _topItems

    init {
        loadMockRevenueData()
    }

    @Suppress("UNCHECKED_CAST")
    private fun loadMockRevenueData() {
        val sources = listOf("Consultation Fees", "Marketplace Commissions", "Subscription Fees", "Premium Content Sales")
        val revenueBySourceMtd = sources.associateWith { Random.nextDouble(1000.0, 10000.0) }
        val revenueBySourceYtd = revenueBySourceMtd.mapValues { it.value * Random.nextDouble(8.0, 12.0) }

        val totalRevenueMtd = revenueBySourceMtd.values.sum()
        val totalRevenueYtd = revenueBySourceYtd.values.sum()

        _summary.value = RevenueSummary(mapOf(
            "total_revenue_mtd" to totalRevenueMtd,
            "total_revenue_ytd" to totalRevenueYtd,
            "average_revenue_per_paying_user_arppu_monthly" to totalRevenueMtd / (1000 + Random.nextInt(-100, 100)),
            "churn_rate_monthly" to "${Random.nextDouble(1.5, 5.0).format(2)}%"
        ))

        _revenueBySource.value = RevenueBySource(mapOf("mtd" to revenueBySourceMtd, "ytd" to revenueBySourceYtd))

        val today = Calendar.getInstance()
        val dailyPoints = (0..29).map { i ->
            val day = today.clone() as Calendar
            day.add(Calendar.DATE, -i)
            RevenueTrendPoint(
                label = "${day.get(Calendar.YEAR)}-${day.get(Calendar.MONTH) + 1}-${day.get(Calendar.DAY_OF_MONTH)}",
                revenue = sources.sumOf { Random.nextDouble(50.0, 200.0) }
            )
        }.reversed()
        _dailyTrend.value = RevenueTrend("Daily", dailyPoints)

        val monthlyPoints = (0..11).map { i ->
            val currentMonthCal = today.clone() as Calendar
            currentMonthCal.add(Calendar.MONTH, -i)
            RevenueTrendPoint(
                label = "${currentMonthCal.get(Calendar.YEAR)}-${currentMonthCal.get(Calendar.MONTH) + 1}",
                revenue = sources.sumOf { Random.nextDouble(5000.0, 20000.0) }
            )
        }.reversed()
        _monthlyTrend.value = RevenueTrend("Monthly", monthlyPoints)

        _clv.value = ClvData(mapOf(
            "farmers_standard" to Random.nextDouble(50.0, 150.0),
            "farmers_premium" to Random.nextDouble(150.0, 400.0),
            "veterinarians_basic" to Random.nextDouble(100.0, 250.0),
            "veterinarians_pro" to Random.nextDouble(250.0, 600.0),
            "overall_average" to (List(4) { Random.nextDouble(50.0, 600.0) }).average()
        ))

        _topItems.value = listOf(
            TopRevenueItem("consult_specialist_A", "Consultation", Random.nextDouble(500.0, 1500.0)),
            TopRevenueItem("product_feed_B", "Marketplace", Random.nextDouble(300.0, 1000.0)),
            TopRevenueItem("subscription_vet_pro", "Subscription", Random.nextDouble(1000.0, 2000.0))
        )
    }
    // Helper extension for formatting
    fun Double.format(digits: Int) = "%.${digits}f".format(this)
}


// --- Composable Screen ---
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun RevenueAnalyticsScreen(viewModel: RevenueAnalyticsViewModel = androidx.lifecycle.viewmodel.compose.viewModel()) {
    val summary by viewModel.summary.collectAsState()
    val revenueBySource by viewModel.revenueBySource.collectAsState()
    val dailyTrend by viewModel.dailyTrend.collectAsState()
    val monthlyTrend by viewModel.monthlyTrend.collectAsState()
    val clv by viewModel.clv.collectAsState()
    val topItems by viewModel.topItems.collectAsState()

    val decimalFormat = remember { DecimalFormat("$ #,##0.00") }

    Scaffold(
        topBar = {
            TopAppBar(
                title = { Text("Revenue Analytics") },
                colors = TopAppBarDefaults.topAppBarColors(containerColor = Color(0xFFAB47BC)) // Purple
            )
        }
    ) { paddingValues ->
        LazyColumn(modifier = Modifier.padding(paddingValues).padding(16.dp)) {
            summary?.let { item { RevenueSectionCard("Revenue Summary", it.metrics, decimalFormat) } }
            revenueBySource?.let {
                item { RevenueSectionCard("Revenue by Source (MTD)", it.periodData["mtd"] ?: emptyMap(), decimalFormat) }
                item { RevenueSectionCard("Revenue by Source (YTD)", it.periodData["ytd"] ?: emptyMap(), decimalFormat) }
            }
            dailyTrend?.let { item { RevenueTrendCard("Daily Revenue Trend (Last 5 Days)", it.points.takeLast(5), decimalFormat) } }
            monthlyTrend?.let { item { RevenueTrendCard("Monthly Revenue Trend (Last 3 Months)", it.points.takeLast(3), decimalFormat) } }
            clv?.let { item { RevenueSectionCard("Customer Lifetime Value (CLV)", it.segments, decimalFormat) } }
            if (topItems.isNotEmpty()) {
                item { TopRevenueItemsCard(topItems, decimalFormat) }
            }
        }
    }
}

@Composable
fun RevenueSectionCard(title: String, metrics: Map<String, Any>, formatter: DecimalFormat) {
    Card(modifier = Modifier.fillMaxWidth().padding(vertical = 8.dp), elevation = CardDefaults.cardElevation(2.dp)) {
        Column(Modifier.padding(16.dp)) {
            Text(title, style = MaterialTheme.typography.headlineSmall, modifier = Modifier.padding(bottom = 8.dp))
            DisplayRevenueMetrics(metrics, formatter)
        }
    }
}

@Composable
fun DisplayRevenueMetrics(metrics: Map<String, Any>, formatter: DecimalFormat, indentLevel: Int = 0) {
    val indent = "  ".repeat(indentLevel)
    metrics.forEach { (key, value) ->
        val displayKey = key.replace("_", " ").replaceFirstChar { if (it.isLowerCase()) it.titlecase() else it.toString() }
        when (value) {
            is Map<*, *> -> {
                Text("$indent$displayKey:")
                DisplayRevenueMetrics(value as Map<String, Any>, formatter, indentLevel + 1)
            }
            is Double -> Text("$indent$displayKey: ${formatter.format(value)}")
            is Number -> Text("$indent$displayKey: $value") // For counts or non-currency numbers
            else -> Text("$indent$displayKey: $value")
        }
    }
}

@Composable
fun RevenueTrendCard(title: String, points: List<RevenueTrendPoint>, formatter: DecimalFormat) {
    Card(modifier = Modifier.fillMaxWidth().padding(vertical = 8.dp), elevation = CardDefaults.cardElevation(2.dp)) {
        Column(Modifier.padding(16.dp)) {
            Text(title, style = MaterialTheme.typography.headlineSmall, modifier = Modifier.padding(bottom = 8.dp))
            points.forEach { point ->
                Text("  ${point.label}: ${formatter.format(point.revenue)}")
            }
            // Placeholder for chart
            // Text("Chart would be here", textAlign = androidx.compose.ui.text.style.TextAlign.Center, modifier = Modifier.fillMaxWidth().padding(top=8.dp))
        }
    }
}

@Composable
fun TopRevenueItemsCard(items: List<TopRevenueItem>, formatter: DecimalFormat) {
    Card(modifier = Modifier.fillMaxWidth().padding(vertical = 8.dp), elevation = CardDefaults.cardElevation(2.dp)) {
        Column(Modifier.padding(16.dp)) {
            Text("Top Revenue Generating Items (MTD)", style = MaterialTheme.typography.headlineSmall, modifier = Modifier.padding(bottom = 8.dp))
            items.forEach { item ->
                Text("  ${item.name.replace("_", " ").titlecase()} (${item.type}): ${formatter.format(item.revenueMtd)}")
            }
        }
    }
}

// String.titlecase() helper if not available or for specific locale
fun String.titlecase(): String = replaceFirstChar { if (it.isLowerCase()) it.titlecase(Locale.getDefault()) else it.toString() }


@Preview(showBackground = true)
@Composable
fun PreviewRevenueAnalyticsScreen() {
    MaterialTheme {
        RevenueAnalyticsScreen(viewModel = RevenueAnalyticsViewModel())
    }
}
